api/v1/discovery_webhook.go

```diff
@@ -0,0 +1,164 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"errors"
+	"fmt"
+
+	admissionregistration "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	cl "sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// log is for logging in this package.
+var (
+	discoveredclusterLog = logf.Log.WithName("discoveredcluster-resource")
+	Client               cl.Client
+
+	ErrInvalidImportStrategy = errors.New("invalid import-strategy")
+)
+
+// ValidatingWebhook returns the ValidatingWebhookConfiguration used for the discoveredcluster
+// linked to a service in the provided namespace
+func ValidatingWebhook(namespace string) *admissionregistration.ValidatingWebhookConfiguration {
+	fail := admissionregistration.Fail
+	none := admissionregistration.SideEffectClassNone
+	path := "/validate-discovery-open-cluster-management-io-v1-discoveredcluster"
+	return &admissionregistration.ValidatingWebhookConfiguration{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "admissionregistration.k8s.io/v1",
+			Kind:       "ValidatingWebhookConfiguration",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "discovery.open-cluster-management.io",
+			Annotations: map[string]string{"service.beta.openshift.io/inject-cabundle": "true"},
+		},
+		Webhooks: []admissionregistration.ValidatingWebhook{
+			{
+				AdmissionReviewVersions: []string{
+					"v1",
+					"v1beta1",
+				},
+				Name: "discovery.open-cluster-management.io",
+				ClientConfig: admissionregistration.WebhookClientConfig{
+					Service: &admissionregistration.ServiceReference{
+						Name:      "discovery-operator-webhook-service",
+						Namespace: namespace,
+						Path:      &path,
+					},
+				},
+				FailurePolicy: &fail,
+				Rules: []admissionregistration.RuleWithOperations{
+					{
+						Rule: admissionregistration.Rule{
+							APIGroups:   []string{GroupVersion.Group},
+							APIVersions: []string{GroupVersion.Version},
+							Resources:   []string{"discoveredclusters"},
+						},
+						Operations: []admissionregistration.OperationType{
+							admissionregistration.Create,
+							admissionregistration.Update,
+							admissionregistration.Delete,
+						},
+					},
+				},
+				SideEffects: &none,
+			},
+		},
+	}
+}
+
+func (r *DiscoveredCluster) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	Client = mgr.GetClient()
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+var _ webhook.Defaulter = &DiscoveredCluster{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *DiscoveredCluster) Default() {
```

Comment:
This is just for setting the `default` setting of the webhook. No harm in just having this log for the moment, since this func can be enhanced at a later time.